### PR TITLE
MET-5968 Country Target Data (all)

### PR DESCRIPTION
**Pull request by andyjmaclean**

### CODE REVIEW

**Review by JoanaCMS — CHANGES_REQUESTED** [review]

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * Class that encapsulates historical data for a specific date
+ */
+public class CountryTargetResult {
+
+    private String country;
+    private String targetType;
+    private String label;
+    private int value;
+    private boolean interim;
+
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java:15]: The front end does not expect the percentage from the back end.

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * @return all HistoricalCountryTargetDataResult objects
+     */
+    public List<CountryTargetResult> getCountryTargets(){
+
+      List<Target> targetData = mongoSDTargetsDao.getCountryTargets();
+
+      ArrayList<CountryTargetResult> result = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java]: Done

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/persistence/MongoSDTargetsDao.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    final MapperOptions mapperOptions = MapperOptions.builder()
+        .collectionNaming(NamingStrategy.identity()).build();
+    this.datastore = Morphia.createDatastore(mongoClient, mongoDatabaseName, mapperOptions);
+  }
+
+  /**
+   * Return all Historical entities
+   */
+  public List<Historical> getCountryTargetData(){
+    ArrayList<Historical> queryResult = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/persistence/MongoSDTargetsDao.java]: Done

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/persistence/MongoSDTargetsDao.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            datastore.find(Historical.class));
+    result.forEach(queryResult::add);
+    return queryResult;
+  }
+
+  /**
+   * Return all Target entities
+   */
+  public List<Target> getCountryTargets(){
+    ArrayList<Target> queryResult = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/persistence/MongoSDTargetsDao.java]: Done

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/HistoricalCountryTargetDataResult.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Class that encapsulates historical data for a specific date
+ */
+public class HistoricalCountryTargetDataResult {
+
+    private LocalDateTime date;
+    private String country;
+
+    private int three_d;
+    private int hq;
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/HistoricalCountryTargetDataResult.java]: Done, but I've left it mapped to 'hq' in the JSON mapping.

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/HistoricalCountryTargetDataResult.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * Class that encapsulates historical data for a specific date
+ */
+public class HistoricalCountryTargetDataResult {
+
+    private LocalDateTime date;
+    private String country;
+
+    private int three_d;
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/HistoricalCountryTargetDataResult.java]: Done.

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Class that encapsulates historical data for a specific date
+ */
+public class CountryTargetResult {
+
+    private String country;
+    private String targetType;
+    private String label;
+    private int value;
+    private boolean interim;
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java]: `interim` flags a target as being the interim target before the 2030 target.  It's used in the UI to decide whether or not the target line in the graph is drawn with a solid style or with a dashed style.  It's a derived field so I've removed it from the API.

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
+package eu.europeana.statistics.dashboard.common.api.response.targetdata;
+
+import java.util.List;
+
+/**
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java:12]: Done

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/HistoricalCountryTargetDataResult.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    public String getCountry() {
+        return country;
+    }
+
+    public void setCountry(String country) {
+        this.country = country;
+    }
+
+    public int getThree_d() {
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/HistoricalCountryTargetDataResult.java]: Done

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/HistoricalCountryTargetDataResult.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        return three_d;
+    }
+
+    public void setTHREE_D(Integer three_d) {
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/HistoricalCountryTargetDataResult.java]: Done

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public String[] getTargetTypes(){
+      return targetTypes;
+    }
+
+    /**
+     * @return all HistoricalCountryTargetDataResult objects
+     */
+    public List<HistoricalCountryTargetDataResult> getCountryDataAll(){
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java]: Done

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    public CountryTargetService(MongoSDTargetsDao mongoSDTargetsDao) {
+        this.mongoSDTargetsDao = mongoSDTargetsDao;
+    }
+
+    public String[] getTargetTypes(){
+      return targetTypes;
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java:31]: Removed

File: statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/controller/CountryTargetDataController.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+     * @return HistoricalCountryTargetDataResult
+     */
+    @GetMapping(value = TARGET_DATA_COUNTRY_ALL, produces = {APPLICATION_JSON})
+    @ResponseStatus(HttpStatus.OK)
+    @ApiOperation(value = "Returns country target data", response = HistoricalCountryTargetDataResult.class)
+
+    public List<HistoricalCountryTargetDataResult> getCountryDataAll(){
+        return countryTargetService.getCountryDataAll();
+    }
+
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/controller/CountryTargetDataController.java:56]: Done

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+     *
+     * @param mongoSDTargetsDao - The mongo database it connects to
+     */
+    public CountryTargetService(MongoSDTargetsDao mongoSDTargetsDao) {
+        this.mongoSDTargetsDao = mongoSDTargetsDao;
+    }
+
+    public String[] getTargetTypes(){
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java]: Removed

File: statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/controller/CountryTargetDataController.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * Autowired constructor
+     */
+    @Autowired
+    public CountryTargetDataController(CountryTargetService countryTargetService) {
+        this.countryTargetService = countryTargetService;
+    }
+
+    /**
+     *
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/controller/CountryTargetDataController.java]: I've refactored the class name "HistoricalCountryTargetData".

How much detail is needed in the Javadoc description from a simple getter?

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+      return historicalData.stream().map(data -> new HistoricalCountryTargetDataResult(
+          data.getCountry(),
+          data.getTimestamp(),
+          data.getThreeD(),
+          data.getHighQuality(),
+          data.getTotalRecords())
+      ).toList();
+    }
+
+    public List<Target> getCountryTargetsRaw(){
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java]: Deleted

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+      return mongoSDTargetsDao.getCountryTargets();
+    }
+
+    /**
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java:45]: Again, how much detail is required for a simple getter with a clear signature?

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+ * Class that encapsulates historical data for a specific date
+ */
+public class CountryTargetResult {
+
+    private String country;
+    private String targetType;
+    private String label;
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java]: It's been removed.

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+     * @return all HistoricalCountryTargetDataResult objects
+     */
+    public List<CountryTargetResult> getCountryTargets(){
+
+      List<Target> targetData = mongoSDTargetsDao.getCountryTargets();
+
+      ArrayList<CountryTargetResult> result = new ArrayList<>();
+
+      for (Target data : targetData) {
+        String label = Integer.toString(data.getYear());
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java]: Done

File: statistics-dashboard-service/src/test/java/eu/europeana/statistics/dashboard/service/CountryTargetServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/test/java/eu/europeana/statistics/dashboard/service/CountryTargetServiceTest.java:17]: Done

File: statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/config/ApplicationConfiguration.java [hunk 21]
```diff
@@ -87,12 +89,22 @@ public MongoSDDao getMongoSDDao(){
     return new MongoSDDao(mongoClient, properties.getMongoDatabaseName(), false);
   }
 
+  @Bean
+  public MongoSDTargetsDao getMongoSDTargetsDao(){
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/config/ApplicationConfiguration.java]: Done

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+public class CountryTargetService {
+
+    private final MongoSDTargetsDao mongoSDTargetsDao;
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java]: Done

File: statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+
+import java.util.List;
+
+/**
+ * Class that encapsulates historical data for a specific date
+ */
+public class CountryTargetResult {
+
+    private String country;
+    private String targetType;
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-common/src/main/java/eu/europeana/statistics/dashboard/common/api/response/targetdata/CountryTargetResult.java]: Done

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+
+      List<Target> targetData = mongoSDTargetsDao.getCountryTargets();
+
+      ArrayList<CountryTargetResult> result = new ArrayList<>();
+
+      for (Target data : targetData) {
+        String label = Integer.toString(data.getYear());
+        String country = data.getCountry();
+        boolean interim = data.getYear() == 2025;
+        int[] values = new int[]{ data.getThreeD(), data.getHighQuality(), data.getTotalRecords() };
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java]: Done

File: statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+
+        for (int i = 0; i < values.length; i++) {
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-service/src/main/java/eu/europeana/statistics/dashboard/service/CountryTargetService.java]: Removed

File: statistics-dashboard-service/src/test/java/eu/europeana/statistics/dashboard/service/CountryTargetServiceTest.java [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+      assertEquals("three_d", result.get(0).getTargetType().getValue());
+      assertEquals("high_quality", result.get(1).getTargetType().getValue());
+      assertEquals("total", result.get(2).getTargetType().getValue());
+
+      for (int i = 0; i < 3; i++) {
+        assertEquals(daoResultValues[i], result.get(i).getValue());
+      }
+   }
+
```
Comment (JoanaCMS) [inline statistics-dashboard-service/src/test/java/eu/europeana/statistics/dashboard/service/CountryTargetServiceTest.java:56]: Missing a test for the method ` getAllCountryData()`

File: statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/controller/CountryTargetDataController.java [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+import io.swagger.annotations.*;
+import io.swagger.v3.oas.annotations.tags.Tag;
+import io.swagger.v3.oas.annotations.tags.Tags;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+/**
+ * Controller for the Target Data Page
+ */
```
Comment (JoanaCMS) [inline statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/controller/CountryTargetDataController.java:19]: Missing a test class for this class

File: statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/config/ApplicationConfiguration.java [hunk 28]
```diff
@@ -87,12 +89,22 @@ public MongoSDDao getMongoSDDao(){
...
+  }
+
   @Bean
   public StatisticsService getStatisticsService(MongoSDDao mongoSDDao){
     this.statisticsService = new StatisticsService(mongoSDDao);
     return statisticsService;
   }
 
+  @Bean
+  public CountryTargetService getCountryTargetService(MongoSDTargetsDao mongoSDTargetsDao){
```
[the remark this answers is not in the document]
Comment (andyjmaclean) [reply statistics-dashboard-rest/src/main/java/eu/europeana/statistics/dashboard/rest/config/ApplicationConfiguration.java]: Done

**Pull request merged**